controller/rest/dlp_rule.go

```diff
@@ -1809,3 +1851,450 @@ func importDlp(scope string, loginDomainRoles access.DomainRole, importTask shar
 
 	return nil
 }
+
+// caller has been verified for federal admin access right
+func replaceFedDlpSensorGroups(DlpSensors []*share.CLUSDlpSensor, DlpGroups []*share.CLUSDlpGroup) bool {
+	dlpSenMap := make(map[string]*share.CLUSDlpSensor, len(DlpSensors))
+	for _, dlpsen := range DlpSensors {
+		if dlpsen.CfgType == share.FederalCfg {
+			dlpSenMap[dlpsen.Name] = dlpsen
+		}
+	}
+	dlpGrpMap := make(map[string]*share.CLUSDlpGroup, len(DlpGroups))
+	for _, dlpgrp := range DlpGroups {
+		if dlpgrp.CfgType == share.FederalCfg {
+			dlpGrpMap[dlpgrp.Name] = dlpgrp
+		}
+	}
+
+	lock, err := clusHelper.AcquireLock(share.CLUSLockPolicyKey, clusterLockWait)
+	if err != nil {
+		log.WithFields(log.Fields{"error": err}).Error("Failed to acquire cluster lock")
+		return false
+	}
+	defer clusHelper.ReleaseLock(lock)
+
+	var defsensor *share.CLUSDlpSensor
+	defsensor = clusHelper.GetDlpSensor(share.CLUSDlpDefaultSensor)
+
+	/*
+	* If the default/predefined dlp sensor is not
+	* created in upgrading process, create it here.
+	 */
+	if defsensor == nil {
+		CreatePredefaultSensor()
+		defsensor = clusHelper.GetDlpSensor(share.CLUSDlpDefaultSensor)
+		if defsensor == nil {
+			e := "sensor cannot be created in cluster!"
+			log.Error(e)
+			return false
+		}
+		log.Debug("Creating predefined sensor!")
+	}
+
+	if defsensor.RuleList == nil {
+		defsensor.RuleList = make(map[string]*share.CLUSDlpRule)
+	}
+	if defsensor.PreRuleList == nil {
+		defsensor.PreRuleList = make(map[string][]*share.CLUSDlpRule)
+	}
+
+	txn := cluster.Transact()
+	defer txn.Close()
+
+	exSensorMap := make(map[string]*share.CLUSDlpSensor)
+	existingSensors := clusHelper.GetAllDlpSensors()
+	for _, esen := range existingSensors {
+		if esen != nil {
+			exSensorMap[esen.Name] = esen
+		}
+		if esen != nil && esen.CfgType == share.FederalCfg {
+			if _, ok := dlpSenMap[esen.Name]; !ok { // in existing but not in new. so delete it
+				for _, rn := range esen.RuleListNames {
+					_, foundInAll := defsensor.RuleList[rn]
+					if foundInAll {
+						delete(defsensor.RuleList, rn)
+					}
+				}
+				clusHelper.DeleteDlpSensorTxn(txn, esen.Name)
+			}
+		}
+	}
+
+	for _, dsen := range dlpSenMap {
+		if dsen != nil {
+			if dsen.Name == share.CLUSFedDlpDefSyncSensor {
+				for rname, cdr := range dsen.RuleList {
+					cdr.ID = getDlpRuleID(defsensor)
```

Comment:
when federal rule is synced to managed cluster, the rule id is regenerated within managed cluster to avoid conflict, this id is only effective at the local cluster.